### PR TITLE
add: error handling to SDK

### LINKED ISSUE #231: add: ability to handle errors on the SDK
### Is your proposal related to a problem?

I don't see currently a way to use a custom error handler for any errors that might occur when submitting the form. For ex: network errors

### Describe the solution you'd like

I actually made a PR (#201) a while back adding this functionality. I think it could be done kinda the same way by having a function on `formbricks.init` that gets called with the error.

### Additional context

This is really a must since the website developer needs to have a way of notifying the end user if anything goes wrong.

### PR DESCRIPTION
**Pull request by kymppi**
Closes #231

### CODE REVIEW

File: packages/js/src/lib/noCodeEvents.ts [hunk 1]
```diff
@@ -40,23 +43,39 @@ export const addPageUrlEventListeners = (): void => {
...
     case "endsWith":
-      return url.endsWith(pageUrlValue);
+      result = url.endsWith(pageUrlValue);
     case "notMatch":
-      return url !== pageUrlValue;
+      result = url !== pageUrlValue;
     case "notContains":
-      return !url.includes(pageUrlValue);
+      result = !url.includes(pageUrlValue);
     default:
```
Comment (mattinannt) [inline packages/js/src/lib/noCodeEvents.ts:67]: case falls through to error case. please add `break;`
(was missing before, I know)

File: packages/js/src/lib/init.ts [hunk 2]
```diff
@@ -24,48 +33,101 @@ const addSessionEventListeners = (): void => {
...
+    const { session, settings } = createSessionResult.value;
+
     config.update({ session: extendSession(session), settings });
-    trackEvent("New Session");
-  } else if (!config.get().session) {
-    logger.error("Unable to initialize. No session found");
+
+    const trackEventResult = await trackEvent("New Session");
+
+    if (trackEventResult.ok !== true) return err(trackEventResult.error);
```
Comment (mattinannt) [inline packages/js/src/lib/init.ts:97]: I'm getting the error
`TypeError: Cannot read properties of undefined (reading 'ok')`

**Review by mattinannt — CHANGES_REQUESTED** [review]
@kymppi thanks a lot for the PR 🤗 Makes total sense to add better error handling and really happy that you tackle this challenge.
I still encounter a few errors while testing. Can you please take a look at them and also rebase / merge the latest changes from main. E.g. I needed to change the init() flow to fix a few problems with one of our clients.
thanks a lot 🤗

**Comment by kymppi** [comment]
**Important*': every function that returns the Result<T, E> type has to return one of these `ok(<value>)` `okVoid()` `err(<error>)`

**Comment by mattinannt** [comment]
@kymppi Thanks for the changes 💪🎉 Looks great :-) 
I found two bugs and fixed them.
And I still have problems using the error handler. I added it to the init function but it's still the default error handler (logger) and I also can't find where it's set in the code.
Screenshot 2023-04-25 at 11 51 45
Also can you provide an object type for the `error` attribute. Since it's `any` I don't know if it's a real error object (with message attribute) or just a string.
If this is fixed I think we are ready to merge 💪

**Comment by kymppi** [comment]
The default errorHandler is defined in config.ts, but currently its hard to add a type to the error since the errors are defined in the js package but the JsConfig and InitConfig types are from the types package

**Comment by kymppi** [comment]
Okay, sorry for not testing it first. I'm currently rewriting how the error handler is initialized and how it is accessed 👍

**Review by mattinannt — APPROVED** [review]
thanks for the changes 💪 works great now 😊🚀🤗
I am going to merge it into main now. So it will go out with the next release probably tomorrow 😊🎉

**Pull request merged**